packages/protocol/contracts/rrp/authorizers/Api3TokenLock.sol

```diff
@@ -0,0 +1,393 @@
+        require(_api3Token != address(0), ERROR_ZERO_ADDRESS);
+        api3Token = _api3Token;
+
+        bool isBurner = IApi3Token(api3Token).getBurnerStatus(address(this));
+        if (!isBurner) {
+            IApi3Token(api3Token).updateBurnerStatus(true);
+        }
+
+        emit SetApi3Token(api3Token);
+    }
+
+    /// @notice Called by owner to set the minimum locking time
+    /// @param _minimumLockingTime The new minimum locking time
+    function setMinimumLockingTime(uint256 _minimumLockingTime)
+        external
+        override
+        onlyOwner
+    {
+        minimumLockingTime = _minimumLockingTime;
+        emit SetMinimumLockingTime(_minimumLockingTime);
+    }
+
+    /// @notice Called by owner to set the locking amount
+    /// @param _lockAmount The new lock amount
+    function setLockAmount(uint256 _lockAmount) external override onlyOwner {
+        require(_lockAmount != 0, ERROR_ZERO_AMOUNT);
+        lockAmount = _lockAmount;
+        emit SetLockAmount(_lockAmount);
+    }
+
+    /// @notice Called by owner to set the blacklist status of a client
+    /// @param clientAddress Client address
+    /// @param status Blacklist status to be set
+    function setBlacklistStatus(address clientAddress, bool status)
+        external
+        override
+        onlyOwner
+    {
+        clientAddressToBlacklistStatus[clientAddress] = status;
+        emit SetBlacklistStatus(clientAddress, status, msg.sender);
+    }
+
+    /// @notice Locks API3 Tokens to gain access to Airnodes.
+    /// Airnode-client pair gets authorized as long as there is at least one token lock for given pair.
+    /// @dev The amount to be locked is determined by a memory variable set by
+    /// the owners of the contract.
+    /// @param _airnodeId The airnode id
+    /// @param _clientAddress The client address
+    function lock(bytes32 _airnodeId, address _clientAddress)
+        external
+        override
+    {
+        AirnodeClient storage airnodeClient = tokenLocks[_airnodeId][
+            _clientAddress
+        ];
+        require(
+            airnodeClient.lockAmountAt[msg.sender] == 0,
+            ERROR_ALREADY_LOCKED
+        );
+        require(
+            !clientAddressToBlacklistStatus[_clientAddress],
+            ERROR_CLIENT_BLACKLISTED
+        );
+
+        uint256 expirationTime = block.timestamp + minimumLockingTime;
+        airnodeClient.lockAmountAt[msg.sender] = lockAmount;
+        airnodeClient.canUnlockAt[msg.sender] = expirationTime;
+        airnodeClient.whitelistCount++;
+
+        assert(
+            IApi3Token(api3Token).transferFrom(
+                msg.sender,
+                address(this),
+                lockAmount
+            )
+        );
+
+        if (airnodeClient.whitelistCount == 1) {
+            IApi3RequesterRrpAuthorizer(api3RequesterRrpAuthorizer)
+                .setWhitelistExpiration(
+                _airnodeId,
+                _clientAddress,
+                type(uint64).max
```

Comment:
What about using `expirationTime` here instead? This way we make sure that authorizer will only authorize the requester while tokens are locked.

Reply:
I don't fully understand what you mean here

Reply:
I was trying to address this question you'd made: 

> Currently the token lock sets the whitelisting expirationTimestamp to type(uint64).max would setting setWhitelistStatusPastExpiration true be better instead?

I personally think it might be a little bit dangerous to set the `whitelistedPastExpiration` var to true and have the requester whitelisted by accident. So my thinking was to remove the check in line [200](https://github.com/api3dao/airnode/blob/57e74a00231a723ecfaedd3c28dea7655227ac88/packages/protocol/contracts/rrp/authorizers/Api3TokenLock.sol#L200) and use the value from [187](https://github.com/api3dao/airnode/blob/57e74a00231a723ecfaedd3c28dea7655227ac88/packages/protocol/contracts/rrp/authorizers/Api3TokenLock.sol#L187) everytime a token lock is added to the airnode-client pair.

Reply:
so the check in line 200 is necessary because of multiple locks of single airnode-requester pair i mentioned before.

Ah i see what you mean. So after the minimumLocking period the requester doesn't need to lock tokens and will be authorized indefinitely according to burak. That's why its been set to infinity.

Reply:
ah! that makes a lot of sense. Thanks for the explanation 💪🏻